### PR TITLE
filter_sdh: optimize get_char_bytes

### PR DESCRIPTION
**Pull request by Dudemanguy**
cc @N-R-K

### CODE REVIEW

File: sub/filter_sdh.c [hunk 1]
```diff
@@ -67,18 +67,22 @@ static inline int append(struct sd_filter *sd, struct buffer *buf, char c)
...
 static int get_char_bytes(char *str)
 {
-    // In case the final character is non-ASCII.
+    // In case the first character is non-ASCII.
     // Will only work with UTF-8 but you shouldn't be
     // using anything else anyway.
     if (str && str[0]) {
         if (!(str[0] >> 7 & 1)) {
-            return MPMIN(strlen(str), 1);
+            int bytes = strnlen(str, 1);
```
Comment (kasper93) [inline sub/filter_sdh.c:75]: No need for MPMIN. `strnlen` will return at most maxlen.

Comment (Dudemanguy) [reply sub/filter_sdh.c]: Duh.

**Review by N-R-K — APPROVED** [review]

**Comment by kasper93** [comment]
In fact there is no need to strnlen every char, `filter_SDH` takes length as an argument. Just need to clamp the value for last one.

**Comment by Dudemanguy** [comment]
Don't think that works since length includes the ASS stuff so it will be longer than `rp`. Also `rp` can get incremented/moved around.

**Comment by N-R-K** [comment]
> Also `rp` can get incremented/moved around.

If you have the length you can compute the "end pointer" which will remain stable and can be used to compute the remaining length. For example:

```c
char *end = start + len;
ptrdiff_t remaining = end - rp;
```

But in any case, my main objection was that it was computing `strlen` every time. `strnlen` - while not as ideal as just passing the length/endpointer - doesn't go through the entire string and so that's fine by me.

**Comment by kasper93** [comment]
My main objection is that `get_char_bytes` does make zero sense to do any strlen. It suppose to read first byte and get length of utf-8 code point. And it did that before e15b2b19a3548c8c210d86a77b52d495494a1e20. If there are truncated code points at the end of the buffer they should be handled at the caller side, it is not `get_char_bytes` job to guess it by doing strlen of each code point in the string.

trivial bound check
```
if (rp >= rp_end)
    ...
```

will achieve the same thing.

**Comment by N-R-K** [comment]
> My main objection is that it make zero sense to do any strlen in `get_char_bytes`.

Yeah, ideally it should. But I'm ok with the current `strnlen` version and don't feel like bike-shedding it.

```
if (rp >= rp_end)
```

FYI, it's undefined behavior for the result of pointer arithmetic to go outside the buffer (one exception is made for 1 past the buffer) [^0]. In other words, compiler is free to assume that the resulting pointer is within bound and make optimization based on that.

So you'd need something like this instead, to do the bound checking *before* adding it (similar to how you need to check for [interger overflow](https://c-faq.com/misc/intovf.html) *before* doing the arithmetic):

```c
n = get_char_bytes(...);
rem = rp_end - rp;
if (n > rem) // OOB
```

[^0]: https://port70.net/~nsz/c/c11/n1570.html#6.5.6p8

**Comment by kasper93** [comment]
> FYI, it's undefined behavior for the result of pointer arithmetic to go outside the buffer

implementation details, I meant we don't need to do strlen to get that information and just check the end.

> But I'm ok with the current strnlen version and don't feel like bike-shedding it.

My initial issue was that we pass-through truncated code points, while it seems like something that should be ignored/fixed. In other words, if first byte tells that the code point length is 4 and we have only 2 bytes remaining it should be handled accordingly and not just silently ignored.

**Comment by N-R-K** [comment]
> not just silently ignored.

Ah I see. I suppose you're correct but I'm not really familiar with what's going on in this file to comment on what the "correct" behavior should be (error out? log and continue?).

I was only interested in this PR since I saw commit https://github.com/mpv-player/mpv/commit/e15b2b19a3548c8c210d86a77b52d495494a1e20 and got alerted by the `strlen` use.

**Comment by Dudemanguy** [comment]
Further improvements are welcome if anyone is up for it, but for now I'll just go with this.

**Pull request merged**